File: crates/contracts/src/request.rs
//! Backtest request contract (backtest-request.schema.json).

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::graph::Graph;
use crate::Decimal;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct PolicySelector {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub profile: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct ExecutionOverrides {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signal_trigger: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub slippage_bps: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gas_model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub action_cooldown: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BacktestConfig {
    pub start: String,
    pub end: String,
    pub interval: String,
    /// venue -> asset -> decimal-string amount
    pub initial_portfolio: BTreeMap<String, BTreeMap<String, Decimal>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub execution: Option<ExecutionOverrides>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BacktestRequest {
    #[serde(default = "default_request_schema_version")]
    pub schema_version: String,
    pub graph: Graph,
    #[serde(default)]
    pub policy: PolicySelector,
    pub config: BacktestConfig,
}

fn default_request_schema_version() -> String {
    "catalyst.backtest.request.v1".to_string()
}

File: packages/contracts/src/catalyst_contracts/request.py
"""Backtest request contract models (backtest-request.schema.json)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from ._base import Decimal, StrictModel
from .graph import Graph
from .policy import PolicyProfile

Interval = Literal["1m", "5m", "15m", "1h", "4h", "1d"]


class PolicySelector(StrictModel):
    """A bare profile selector; a fully resolved policy may also be supplied."""

    profile: PolicyProfile = "strict_v1"


class ExecutionOverrides(StrictModel):
    signal_trigger: (
        Literal["level", "crossing", "crossing_with_cooldown", "once_per_backtest"] | None
    ) = None
    slippage_bps: Decimal | None = None
    gas_model: (
        Literal["none", "fixed_usd", "fixed_native", "historical_fee_history", "historical"] | None
    ) = None
    action_cooldown: str | None = None


class BacktestConfig(StrictModel):
    start: datetime
    end: datetime
    interval: Interval
    # venue -> asset -> decimal-string amount
    initial_portfolio: dict[str, dict[str, Decimal]]
    execution: ExecutionOverrides | None = None


class BacktestRequest(StrictModel):
    schema_version: str = "catalyst.backtest.request.v1"
    graph: Graph
    policy: PolicySelector = Field(default_factory=PolicySelector)
    config: BacktestConfig


__all__ = [
    "BacktestRequest",
    "BacktestConfig",
    "PolicySelector",
    "ExecutionOverrides",
    "Interval",
]
